File: jsync_js/packages/jsync/src/message/pool.ts
import type { PathSegment } from './action.js';

/** Producer-side path segment pool with stable indexes and O(1) segment lookup. */
export class ProducerPathSegmentPool {
  readonly #segments: PathSegment[];
  readonly #indexes = new Map<string, number>();

  constructor(segments: readonly PathSegment[] = []) {
    this.#segments = [...segments];
    this.#segments.forEach((segment, index) => {
      this.#indexes.set(segmentKey(segment), index);
    });
  }

  clone(): ProducerPathSegmentPool {
    return new ProducerPathSegmentPool(this.#segments);
  }

  /** @internal */
  indexOf(segment: PathSegment): number | undefined {
    return this.#indexes.get(segmentKey(segment));
  }

  /** @internal Returns the committed pool size for producer-side cost simulation. */
  get size(): number {
    return this.#segments.length;
  }

  withTransaction<T>(callback: (transaction: ProducerPathSegmentPoolTransaction) => T): T {
    const checkpoint = this.#segments.length;
    const transaction = new ProducerPathSegmentPoolTransaction(
      (path) => path.map((segment) => this.#indexFor(segment)),
      () => this.#appendedSince(checkpoint),
    );
    try {
      const result = callback(transaction);
      if (transaction.aborted) {
        this.#rollbackTo(checkpoint);
      }
      return result;
    } catch (error: unknown) {
      this.#rollbackTo(checkpoint);
      throw error;
    }
  }

  #rollbackTo(length: number): void {
    if (length >= this.#segments.length) return;
    const removed = this.#segments.splice(length);
    for (const segment of removed) {
      this.#indexes.delete(segmentKey(segment));
    }
  }

  #appendedSince(length: number): PathSegment[] {
    return this.#segments.slice(length);
  }

  #indexFor(segment: PathSegment): number {
    const key = segmentKey(segment);
    const existing = this.#indexes.get(key);
    if (existing !== undefined) return existing;

    const index = this.#segments.length;
    this.#segments.push(segment);
    this.#indexes.set(key, index);
    return index;
  }
}

/** Producer-side path segment pool transaction. */
export class ProducerPathSegmentPoolTransaction {
  readonly #encodePath: (path: readonly PathSegment[]) => number[];
  readonly #appendedSegments: () => PathSegment[];
  #aborted = false;

  constructor(
    encodePath: (path: readonly PathSegment[]) => number[],
    appendedSegments: () => PathSegment[],
  ) {
    this.#encodePath = encodePath;
    this.#appendedSegments = appendedSegments;
  }

  encodePath(path: readonly PathSegment[]): number[] {
    return this.#encodePath(path);
  }

  appendedSegments(): PathSegment[] {
    return this.#appendedSegments();
  }

  abort(): void {
    this.#aborted = true;
  }

  get aborted(): boolean {
    return this.#aborted;
  }
}

/** Consumer-side path segment pool with stable indexes. */
export class ConsumerPathSegmentPool {
  readonly #segments: PathSegment[];

  constructor(segments: readonly PathSegment[] = []) {
    this.#segments = [...segments];
  }

  withTransaction<T>(callback: (transaction: ConsumerPathSegmentPoolTransaction) => T): T {
    const checkpoint = this.#segments.length;
    const transaction = new ConsumerPathSegmentPoolTransaction(
      (segments) => this.#appendSegments(segments),
      (index) => this.#pathSegmentAt(index),
      () => this.#segments.length,
    );
    try {
      const result = callback(transaction);
      if (transaction.aborted) {
        this.#rollbackTo(checkpoint);
      }
      return result;
    } catch (error: unknown) {
      this.#rollbackTo(checkpoint);
      throw error;
    }
  }

  #appendSegments(segments: PathSegment[]): void {
    this.#segments.push(...segments);
  }

  #pathSegmentAt(index: number): PathSegment | undefined {
    return this.#segments[index];
  }

  #rollbackTo(length: number): void {
    this.#segments.length = length;
  }
}

/** Consumer-side path segment pool transaction. */
export class ConsumerPathSegmentPoolTransaction {
  readonly #appendSegments: (segments: PathSegment[]) => void;
  readonly #pathSegmentAt: (index: number) => PathSegment | undefined;
  readonly #poolLength: () => number;
  #aborted = false;

  constructor(
    appendSegments: (segments: PathSegment[]) => void,
    pathSegmentAt: (index: number) => PathSegment | undefined,
    poolLength: () => number,
  ) {
    this.#appendSegments = appendSegments;
    this.#pathSegmentAt = pathSegmentAt;
    this.#poolLength = poolLength;
  }

  appendSegments(segments: PathSegment[]): void {
    this.#appendSegments(segments);
  }

  /** @internal */
  pathSegmentAt(index: number): PathSegment | undefined {
    return this.#pathSegmentAt(index);
  }

  /** @internal */
  poolLength(): number {
    return this.#poolLength();
  }

  abort(): void {
    this.#aborted = true;
  }

  get aborted(): boolean {
    return this.#aborted;
  }
}

export function segmentKey(segment: PathSegment): string {
  return typeof segment === 'string' ? `s:${segment}` : `i:${segment}`;
}

File: jsync_rs/crates/jsync_rs/src/message/pool.rs
use std::collections::HashMap;

use super::PathSegment;

/// Producer-side path segment pool with stable indexes and O(1) segment lookup.
#[derive(Debug, Clone, Default)]
pub struct ProducerPathSegmentPool {
    segments: Vec<PathSegment>,
    indexes: HashMap<PathSegment, usize>,
}

impl ProducerPathSegmentPool {
    /// Creates an empty producer path segment pool.
    pub fn new() -> Self {
        Self::default()
    }

    /// Starts an atomic producer pool update.
    pub fn transaction(&mut self) -> ProducerPathSegmentPoolTransaction<'_> {
        let checkpoint = self.segments.len();
        ProducerPathSegmentPoolTransaction {
            pool: self,
            checkpoint,
            committed: false,
        }
    }

    pub(super) fn index_for(&mut self, segment: &PathSegment) -> usize {
        if let Some(index) = self.indexes.get(segment) {
            return *index;
        }

        let index = self.segments.len();
        let segment = segment.clone();
        self.segments.push(segment.clone());
        self.indexes.insert(segment, index);
        index
    }

    pub(crate) fn index_of(&self, segment: &PathSegment) -> Option<usize> {
        self.indexes.get(segment).copied()
    }

    /// Returns the committed pool size so producer-side estimators can simulate
    /// future appended segment indexes without mutating the real pool.
    pub(crate) fn len(&self) -> usize {
        self.segments.len()
    }

    fn rollback_to(&mut self, len: usize) {
        if len >= self.segments.len() {
            return;
        }

        for segment in self.segments.drain(len..) {
            self.indexes.remove(&segment);
        }
    }
}

/// Producer-side path segment pool transaction.
#[derive(Debug)]
pub struct ProducerPathSegmentPoolTransaction<'a> {
    pub(super) pool: &'a mut ProducerPathSegmentPool,
    checkpoint: usize,
    committed: bool,
}

impl ProducerPathSegmentPoolTransaction<'_> {
    /// Returns the segments appended since this transaction started.
    pub fn appended_segments(&self) -> &[PathSegment] {
        &self.pool.segments[self.checkpoint..]
    }

    /// Commits this transaction.
    pub fn commit(mut self) {
        self.committed = true;
    }

    /// Aborts this transaction and rolls the pool back to its checkpoint.
    pub fn abort(self) {}
}

impl Drop for ProducerPathSegmentPoolTransaction<'_> {
    fn drop(&mut self) {
        if !self.committed {
            self.pool.rollback_to(self.checkpoint);
        }
    }
}

/// Consumer-side path segment pool with stable indexes.
#[derive(Debug, Clone, Default)]
pub struct ConsumerPathSegmentPool {
    segments: Vec<PathSegment>,
}

impl ConsumerPathSegmentPool {
    /// Creates an empty consumer path segment pool.
    pub fn new() -> Self {
        Self::default()
    }

    /// Starts an atomic consumer pool update.
    pub fn transaction(&mut self) -> ConsumerPathSegmentPoolTransaction<'_> {
        let checkpoint = self.segments.len();
        ConsumerPathSegmentPoolTransaction {
            pool: self,
            checkpoint,
            committed: false,
        }
    }

    fn rollback_to(&mut self, len: usize) {
        self.segments.truncate(len);
    }
}

/// Consumer-side path segment pool transaction.
#[derive(Debug)]
pub struct ConsumerPathSegmentPoolTransaction<'a> {
    pool: &'a mut ConsumerPathSegmentPool,
    checkpoint: usize,
    committed: bool,
}

impl ConsumerPathSegmentPoolTransaction<'_> {
    /// Appends path segments declared by message metadata.
    pub fn append_segments(&mut self, segments: Vec<PathSegment>) {
        self.pool.segments.extend(segments);
    }

    pub(super) fn path_segment_at(&self, index: usize) -> Option<PathSegment> {
        self.pool.segments.get(index).cloned()
    }

    pub(super) fn pool_len(&self) -> usize {
        self.pool.segments.len()
    }

    /// Commits this transaction.
    pub fn commit(mut self) {
        self.committed = true;
    }

    /// Aborts this transaction and rolls the pool back to its checkpoint.
    pub fn abort(self) {}
}

impl Drop for ConsumerPathSegmentPoolTransaction<'_> {
    fn drop(&mut self) {
        if !self.committed {
            self.pool.rollback_to(self.checkpoint);
        }
    }
}
